File: buildTreefrompost_inorder.py
# Time Complexity : O(N) 
# Space Complexity :    O(N) 
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO



# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if len(inorder)==0 or len(postorder)==0:
            return None
        
        self.map = {}
        self.index = len(postorder)-1 #index in postorder
        
        for i in range(len(inorder)):
            self.map[inorder[i]]=i
            
        return self.helper(postorder,0,len(inorder)-1)
    
    def helper(self,postorder,start,end):
        if start>end :
            return 
        
        rVal = postorder[self.index]
        root = TreeNode(rVal)
        self.index-=1
        inorderIdx = self.map[rVal]
        
        root.right = self.helper(postorder,inorderIdx+1,end)
        root.left = self.helper(postorder,start,inorderIdx-1)
        return root


        

File: BuildTreeFromPostInOrder.java
// Time Complexity : O(n) where n is number of node in the binary tree
// Space Complexity : O(h)+ O(n) where h is height of binary tree. 
//                  in worst case the height of binary tree can be n, so space can be O(n) in that case. 
//                  And n space is used to store inorder array index mapping
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

import java.util.*;
public class BuildTreeFromPostInOrder {
    public static void main(String[] args) {
        
    }
}

class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(inorder == null || inorder.length == 0 || postorder == null || postorder.length == 0){
            return null;
        }
        int len = inorder.length;
        Map<Integer, Integer> map = new HashMap<>();
        
        for(int i=0; i<len; i++){
            map.put(inorder[i], i);
        }
        int[] idx = new int[1];
        idx[0] = len-1;
        return helper(inorder, postorder,0, len-1, idx, map);
    }
    private TreeNode helper(int[] inorder, int[] postorder, int start, int end, int[] idx, Map<Integer, Integer> map){
        
        //base
        if(start > end){
            return null;
        }
        
        //logic
        //get root node value
        int rootVal = postorder[idx[0]--];
        TreeNode root = new TreeNode(rootVal);
        
        //find root position in inorder
        int rootPos = map.get(rootVal);
        root.right = helper(inorder, postorder, rootPos+1, end, idx, map);
        root.left = helper(inorder, postorder, start, rootPos-1, idx, map);
        
        return root;
    }
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
